src/index.cpp

```diff
@@ -1545,37 +1554,43 @@ namespace diskann {
     ScratchStoreManager<InMemQueryScratch<T>> manager(_query_scratch);
     auto                                      scratch = manager.scratch_space();
 
-    if (L > scratch->search_l) {
-      scratch->resize_for_query(L);
-      diskann::cout << "Expanding query scratch_space. Was created with Lsize: "
-                    << scratch->search_l << " but search L is: " << L
-                    << std::endl;
+    if (L > scratch->get_L()) {
+      diskann::cout << "Attempting to expand query scratch_space. Was created "
+                    << "with Lsize: " << scratch->get_L()
+                    << " but search L is: " << L << std::endl;
+      scratch->resize_for_new_L(L);
+      diskann::cout << "Resize completed. Let scratch->L is "
+                    << scratch->get_L() << std::endl;
     }
-    _u32  *indices = scratch->indices();
-    float *dist_interim = scratch->interim_dists();
-    search_impl(query, L, L, indices, dist_interim, scratch);
 
     std::shared_lock<std::shared_timed_mutex> ul(_update_lock);
+
+    std::vector<unsigned> init_ids(1, _start);
+    iterate_to_fixed_point(query, L, init_ids, scratch, true, true);
+    NeighborPriorityQueue &best_L_nodes = scratch->best_l_nodes();
+    assert(best_L_nodes.size() <= L);
+
     std::shared_lock<std::shared_timed_mutex> tl(_tag_lock);
-    size_t                                    pos = 0;
 
+    size_t pos = 0;
     for (int i = 0; i < (int) L; ++i) {
       TagT tag;
+      auto node = best_L_nodes[i];
```

Comment:
We don't know that there are `L` entries in `best_L_nodes`, could be less. And we seem to be using `i` just for accessing that vector.

How about changing the `for` loop to say:
```
for (const auto& node : best_L_nodes) {
```

That way we only iterate over what's actually inside `best_L_nodes`.

Reply:
You are right. Changed it to stop at best_L_nodes.size()